workflow/util/util_test.go

```diff
@@ -2102,3 +2102,367 @@ func TestRetryWorkflowWithNestedDAGsWithSuspendNodes(t *testing.T) {
+          - failure
+          name: task-after-failure
+          template: step
+        - arguments:
+            parameters:
+            - name: exitCode
+              value: "2"
+          continueOn:
+            failed: true
+          dependencies:
+          - success
+          name: continue
+          template: step
+        - arguments:
+            parameters:
+            - name: exitCode
+              value: "0"
+          dependencies:
+          - continue
+          name: task-after-continue
+          template: step
+      inputs: {}
+      metadata: {}
+      name: dag
+      outputs: {}
+    namespaced/dag-to-retry/step:
+      container:
+        command:
+        - sh
+        - -c
+        - exit {{inputs.parameters.exitCode}}
+        image: alpine:3.7
+        name: ""
+        resources: {}
+      inputs:
+        parameters:
+        - name: exitCode
+      metadata: {}
+      name: step
+      outputs: {}
+  storedWorkflowTemplateSpec:
+    activeDeadlineSeconds: 300
+    arguments: {}
+    entrypoint: dag
+    podMetadata:
+      annotations:
+        sidecar.istio.io/inject: "false"
+    podSpecPatch: |
+      terminationGracePeriodSeconds: 3
+    templates:
+    - container:
+        command:
+        - sh
+        - -c
+        - exit {{inputs.parameters.exitCode}}
+        image: alpine:3.7
+        name: ""
+        resources: {}
+      inputs:
+        parameters:
+        - name: exitCode
+      metadata: {}
+      name: step
+      outputs: {}
+    - dag:
+        failFast: false
+        tasks:
+        - arguments:
+            parameters:
+            - name: exitCode
+              value: "0"
+          name: success
+          template: step
+        - arguments:
+            parameters:
+            - name: exitCode
+              value: "1"
+          dependencies:
+          - success
+          name: failure
+          template: step
+        - arguments:
+            parameters:
+            - name: exitCode
+              value: "0"
+          dependencies:
+          - failure
+          name: task-after-failure
+          template: step
+        - arguments:
+            parameters:
+            - name: exitCode
+              value: "2"
+          continueOn:
+            failed: true
+          dependencies:
+          - success
+          name: continue
+          template: step
+        - arguments:
+            parameters:
+            - name: exitCode
+              value: "0"
+          dependencies:
+          - continue
+          name: task-after-continue
+          template: step
+      inputs: {}
+      metadata: {}
+      name: dag
+      outputs: {}
+    workflowTemplateRef:
+      name: dag-to-retry
+  taskResultsCompletionStatus:
+    dag-to-retry-tb7r7-1325528633: true
+    dag-to-retry-tb7r7-1670055836: true
+    dag-to-retry-tb7r7-1819567448: true
+    dag-to-retry-tb7r7-1987291523: true`
+
+func TestRetryWorkflowWithContinurOnSucceededNodes(t *testing.T) {
```

Comment:
There's a existing TestFormulateRetryWorkflow test with many different scenarios,
we should refactor this and add to it https://github.com/argoproj/argo-workflows/blob/5ca8b69fe7d8e890d1436ee5bac7dbd87ff10a88/workflow/util/util_test.go#L1340-L1366

Reply:
My thinking is that,
for real workflow, we already have e2e test to handle it,
and here we use a simpler form more unit test-like
> In TestFormulateRetryWorkflow, The wf are all like this:

I think this works. Let me know what you think and if you need assistance